src/openms/include/OpenMS/ANALYSIS/TOPDOWN/DeconvolutedSpectrum.h

```diff
@@ -0,0 +1,144 @@
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/KERNEL/MSSpectrum.h>
+#include <OpenMS/ANALYSIS/TOPDOWN/FLASHDeconvHelperStructs.h>
+#include <OpenMS/ANALYSIS/TOPDOWN/PeakGroup.h>
+#include <iomanip>
+
+namespace OpenMS
+{
+  class PeakGroup;
+
+  /**
+       @brief A class representing a deconvoluted spectrum. Also contains deconvoluted precursro information for MSn n>1.
+  */
+  class OPENMS_DLLAPI DeconvolutedSpectrum :
+      public std::vector<PeakGroup>
+  {
+  public:
+    typedef FLASHDeconvHelperStructs::LogMzPeak LogMzPeak;
+
+    /// default constructor
+    DeconvolutedSpectrum() = default;
+
+    /**
+       @brief Constructor for DeconvolutedSpectrum
+       @param s spectrum
+       @param n scan number
+  */
+    explicit DeconvolutedSpectrum(const MSSpectrum &s, int n);
+
+    /// default deconstructor
+    ~DeconvolutedSpectrum() = default;
+
+    /// copy constructor
+    DeconvolutedSpectrum(const DeconvolutedSpectrum &) = default;
+
+    /// move constructor
+    DeconvolutedSpectrum(DeconvolutedSpectrum &&other) = default;
+
+    /// assignment operator
+    DeconvolutedSpectrum &operator=(const DeconvolutedSpectrum &fd) = default;
+
+    /**
+        @brief write the header in the output file (spectrum level)
+        @param fs file stream to the output file
+        @param n the index to the mass
+        @param detail if set true, detailed information of the mass (e.g., peak list for the mass) is written
+   */
+    static void writeDeconvolutedMassesHeader(std::fstream &fs, int &n, bool detail);
+
+    /**
+      @brief write the deconvoluted masses in the output file (spectrum level)
+      @param fs file stream to the output file
+      @param param FLASHDeconv paramter
+    */
+    void writeDeconvolutedMasses(std::fstream &fs,
+                                 const String &fileName,
+                                 const FLASHDeconvHelperStructs::PrecalculatedAveragine &avg,
+                                 bool writeDetail);
+
+    /**
+      @brief write the deconvoluted masses TopFD format
+      @param fs file stream to the output file
+      @param id the index to the spectrum. updated outside.
+ */
+    void writeTopFD(std::fstream &fs, int id, const FLASHDeconvHelperStructs::PrecalculatedAveragine &avg);
+
+    /// cast DeconvolutedSpectrum into MSSpectrum object to write mzml format
+    MSSpectrum toSpectrum(int charge);
+
+    /// write the header for Thermo Inclusion List header format
+    static void writeThermoInclusionHeader(std::fstream &fs);
+
+    /// for memory save... clear unnecessary information in mass tracing
+    void clearPeakGroupsChargeInfo();
+
+    /**
+     @brief register the precusor info in this from the precursor DeconvolutedSpectrum
+     @param precursorSpectrum the precursor DeconvolutedSpectrum
+     */
+    bool registerPrecursor(DeconvolutedSpectrum &precursorSpectrum);
```

Comment:
- can precursorSpectrum be const?
- rename precursorSpectrum to precursor_spectrum